boards: arm: Add Renesas Spider board

aaillet opened the pull request:
This pull request is introducing Generation 4 of Renesas R-Car SoCs with the S4 SoC as well as the Spider board into Zephyr.

Gen4 SoCs vs Gen3 (Zephyr POV) :

Cortex-R52 is now used as Real Time core
We are now able to use embedded ARM timer
Some IPs are the same and will use existing drivers, sometime with few improvements (UART, GPIO...)
Introduction of new IPs
S4 SoC that is equipping multiple Gen4 boards have some Cortex-A55s and a Cortex-R52 core.

Spider board is the reference board for S4 SoC, developpment oriented, this board is providing a lot of available IOs.

Operation of basic samples such as hello_world and blinky have been validated.

aaillet committed:
dts: arm: Add Renesas r8a779f0 support

r8a779f0 is also know as S4, this SoC is part of the Gen4 SoC series,
has 8 Cortex-A55 and a dual core lockstep Cortex-R52 processor.

SCIF0 is dedicated to Zephyr and SCIF3 to Linux.

**Control Domains**
IMPORTANT: This SoC is divided into two "domains":
- Application domain contains some peripherals as well as A55 & R52 cores.
- Control domain that contain a G4MH/RH850 MCU and other peripherals.

In order to access control domain peripherals such as gpio4-7 and CAN-FD
from application domain, the G4MH MCU has to unlock a protection
mechanism from control domain buses.

"Protected" controllers will be flagged in gen4 device trees,
warning users that they need to flash a custom G4MH firmware
to unlock access to these controllers.

**Clock controller**
This SoC clock controller is offering "domains"
for each world (Zephyr/Linux).

These domains are several "entry points" to the clock controller
which are arbitrated to avoid a world from turning off a clock needed
by another one.

We decided to use the same domain as Linux because the
security mechanism as to be implemented before accessing
another domain.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
boards: arm: Add Renesas Spider support

Renesas Spider board use a S4 SoC (r8a779f0).

Add basic support for UART, GPIO and clock control.

We are using SCIF0 as SCIF3 is used by Linux.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
doc: Add Renesas Spider board documentation

Adding Spider board documentation based on Renesas official documentation
and following Zephyr guideline.

The documentation is describing the board and its current
Zephyr support.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
drivers: pinctrl: rcar: Add r8a779f0 support

Enable PFC controller for r8a779f0 SoC.
Declare pin list for r8a779f0 SoC.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
soc: arm: reneas: Add r8a779f0 support

r8a779f0 SoC is part of the Renesas R-Car Gen4 SoC series.

This SoC has a dual core lockstep Cortex-R52 CPU.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet commented:
This PR is ready to be reviewed as th last CI failures is a false positive !
`pinctrl_soc_pin_t` has to be defined as documented here : https://docs.zephyrproject.org/latest/hardware/pinctrl/index.html#pin-control-drivers.

MaureenHelm commented:
@gmarull @povergoing  can you take a look please?

povergoing reviewed (dismissed):
LGTM for the R52 part

gmarull commented on soc/arm/renesas_rcar/common/pinctrl_rcar.h:
```diff
@@ -0,0 +1,127 @@
+	const uint16_t pins[32];
+};
+
+const struct pfc_bias_reg *pfc_rcar_get_bias_regs(void);
+const struct pfc_drive_reg *pfc_rcar_get_drive_regs(void);
+
+/**
+ * @brief set the register index for a given pin
+ *
+ * @param the pin
+ * @param pointer for the resulting register index
+ * @return 0 if the register index is found, negative
+ * errno otherwise.
+ */
+int pfc_rcar_get_reg_index(uint8_t pin, uint8_t *reg_index);
```
this should not be part of pinctrl_soc, as it's something needed internally by drivers. please, create a new internal header so that drivers/pinctrl.h is not polluted with unnecessary stuff.

aaillet replied:
Hi @gmarull ,
Thanks for your comment.
If my understanding is fine, you want me to move these common declaration to a new file here `zephyr/include/zephyr/drivers/pinctrl/pinctrl_rcar_common.h` ?

aaillet replied:
Hi @gmarull, I'm currently reworking this PR to merge some changes with #60252 , can you confirm that I understood well your remark before I proceed ? Thanks

gmarull replied:
What I mean is that this should only be exposed to code requiring it. An internal header is fine, but decoupled from pinctrl_soc.h

aaillet replied:
Fixed it, please tel me if it's ok now !

henrikbrixandersen reviewed (commented):
GPIO changes okay.

xakep-amatop commented on soc/arm/renesas_rcar/common/pinctrl_rcar.h:
```diff
@@ -0,0 +1,127 @@
+	{								       \
+		.pin = DT_PROP_BY_IDX(node_id, pin, 0),			       \
+		.func = COND_CODE_1(RCAR_HAS_IPSR(node_id),		       \
+				    (RCAR_PIN_FUNC(node_id)), (0)),	       \
```
pls, use struct initializer `{0}` for this field

aaillet replied:
Fixed :+1:

aaillet committed:
drivers: gpio: rcar: Add R-Car Gen4 support

Renesas Gen4 SoCs GPIO IPs are using one more
register comparing to Gen3 SoCs.

The new "INEN" register is used to enable general input.

Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet committed:
drivers: pinctrl: Add R-Car Gen4 support

Renesas R-Car Gen4 is different from Gen3 regarding pinmux.

While Gen3 had only one base address to manage all pins,
Gen4 has one set of pinmux registers per GPIO banks.

We could expose one pinmux register per GPIO controllers,
but that would break potential compatibility with Linux
Device tree.

Instead create a reg_base array to parse all reg base from
device tree and identify proper base address based on the pin
definition.

This imply to add a pfc_base parameter to most of the pfc_rcar
function.

Signed-off-by: Julien Massot <julien.massot@iot.bzh>
Signed-off-by: Pierre Marzin <pierre.marzin@iot.bzh>
Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

xakep-amatop committed:
drivers: clock: rcar: Add r8a779f0 support

Add support of r8a779f0 cpg driver.
r8a779f0 soc has its own clock tree.
Gen4 SoCs common registers addresses have been added in header.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Signed-off-by: Aymeric Aillet <aymeric.aillet@iot.bzh>

aaillet commented:
Just updated the PR with the following changes:
- Rebased on main
- Removed NMI init in soc.c following #59869
- Moved GIC version to DT following #58035
- Updated clock driver following #57172 thanks @xakep-amatop for the help
- Updated pinctrl implementation for Gen4 SoCs

aaillet commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -217,10 +252,13 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
 __boot_func static int pfc_rcar_driver_init(void)
 {
 	DEVICE_MMIO_TOPLEVEL_MAP(pfc, K_MEM_CACHE_NONE);
+	reg_base[0] = DEVICE_MMIO_TOPLEVEL_GET(pfc);
```
thanks, fixed !

carlocaione commented on drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:
```diff
@@ -0,0 +1,207 @@
+{
+	int ret = -ENOTSUP;
+
+	switch (module) {
+	default:
+		break;
+	}
```
why?

aaillet replied:
At the moment, none of the clock used by arm32 support has to be enable here !
The driver is however templated on r8a7795 driver and prepared for arm64 support comming directly after this PR (see #60252)

carlocaione commented on soc/arm/renesas_rcar/gen4/soc.c:
```diff
@@ -0,0 +1,23 @@
+
+	/* Invalidate instruction cache and flush branch target cache */
+	__set_ICIALLU(0);
+	__DSB();
+	__ISB();
```
use the barriers API, see `include/zephyr/sys/barrier.h`

aaillet replied:
Thanks, fixed

carlocaione commented on drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:
```diff
@@ -0,0 +1,207 @@
+
+	enable = !!enable;
+
+	switch (clk_info->module) {
+	default:
+		ret = -ENOTSUP;
+		break;
+	}
```
this is odd.

aaillet replied:
same as for last question, for now, not any clocks used by arm32 support has to be enable manually.
however, this will be used as soon as arm64 support will come up.
These methods are alos needed to be filled into `rcar_cpg_mssr_data`
Would it be enough to add a comment in the driver explaining that waiting for this list to be filled ?

carlocaione replied:
yeah, otherwise this is going to confuse a lot of reviewers :smile:

aaillet replied:
proposed a fix, commenting these part of code, let me know if it can be enough, waiting for arm64 side PR (adding some clocks to these switch case) to be merged !

aaillet commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -10,13 +10,24 @@
+#elif defined(CONFIG_SOC_SERIES_RCAR_GEN4)
+/* swap both arguments */
+#define PFC_REG_ADDRESS(idx, node_id) DT_REG_ADDR_BY_IDX(node_id, idx)
+static const uint32_t reg_base[] = {
```
Thanks for your comment, fixed

aaillet replied:
Fixed, thanks

aaillet commented:
Updated the PR moving pinctrl private function prototype to a new internal header.

gmarull commented on drivers/pinctrl/pfc_rcar.h:
```diff
@@ -0,0 +1,23 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DRIVERS_PINCTRL_PFC_RCAR_H_
+#define ZEPHYR_DRIVERS_PINCTRL_PFC_RCAR_H_
+
+const struct pfc_bias_reg *pfc_rcar_get_bias_regs(void);
+const struct pfc_drive_reg *pfc_rcar_get_drive_regs(void);
+
+/**
+ * @brief set the register index for a given pin
+ *
+ * @param the pin
+ * @param pointer for the resulting register index
+ * @return 0 if the register index is found, negative
+ * errno otherwise.
+ */
+int pfc_rcar_get_reg_index(uint8_t pin, uint8_t *reg_index);
```
this header is not self-contained

aaillet replied:
Sorry, I was not aware of this rule

So, I moved the "pfc_rcar.h" include to the top of pfc_rcar.c includes as proposed in #41543 , it also allows me to test self-containment of this new header.

For it to be fully self-contained, I think I need to include pinctrl.soc that is defining both `pfc_bias_reg` and `pfc_drive_reg`  struct.
I'm asking this because, including stdint is enough for it to build (it seems like gcc is not checking struct def at build) but it seems wrong to me.

Researches led me to this paper[1] explaining that all "incomplete" types would be managed the same at build.

What would be Zephyr guideline in this case ?

[1] : https://beej.us/guide/bgc/html/split/incomplete-types.html

klogg commented:
Hi @aaillet
Do you plan to push this? Seems stalled...

aaillet commented:
> Hi @aaillet Do you plan to push this? Seems stalled...

Yes I plan to.
Not stalled, waiting for reviews, I will tag it to be talked about in a dev meeting soon.

aaillet commented:
- Rebased on main
- Added gpio key field to board dts as introduced in #61115 
Trying to fix CI

aaillet commented:
- Rebased on main
- Disabled L1 cache (introducing issues with reading registers values being cached)

xakep-amatop commented:
> * Rebased on main

Note for the future: the Zephyr documentation recommends performing a push rebase and making your changes in separate updates [Workflow suggestions that help reviewers](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers). 

>     * Disabled L1 cache (introducing issues with reading registers values being cached)

Can you share some details about where you performed disabling and which registers you encountered problems with?

aaillet commented:
> > * Rebased on main
> 
> Note for the future: the Zephyr documentation recommends performing a push rebase and making your changes in separate updates [Workflow suggestions that help reviewers](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers).
> 
Thanks for the link, I'll do it that way next time :+1: 
> > ```
> > * Disabled L1 cache (introducing issues with reading registers values being cached)
> > ```
> 
> Can you share some details about where you performed disabling and which registers you encountered problems with?

Here is how we activated cache for gen4 at first (same way it it done for gen3 socs) : https://github.com/iotbzh/zephyr/commit/96a79f89ae831d56747c09f735a75b4daf26b6df

We encountered problem while working on IPC driver. 
Zephyr was not able to read updated master status through libmetal (it was confirmed that ra.
Even if linux was updating the value of this shared register, Zephyr was still reading the old value.
It was fixed instantly by disabling cache.

I removed it of this PR to avoid contagion to other drivers we might not have seen.
We will soon investigate on this topic

aaillet commented:
Removed a comment in board dts (was not respecting compliance), was useless.

aaillet commented:
@lorc @gmarull @povergoing can you please take another look ?

lorc commented on dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:
```diff
@@ -0,0 +1,70 @@
+			clocks = <&cpg CPG_MOD 702>, <&cpg CPG_CORE R8A779F0_CLK_S0D12_PER>;
+		};
+
+		/* Linux console */
```
AFAIK, Cortex A cluster uses HSCIF0 nowadays.

lorc reviewed (commented):
I have a small comment regarding SCIF3/HSCIF0 used by Cortex A cluster. If I remember correctly, Renesas BSP moved Cortex A to HSCIF0. At least this is what we use in OP-TEE, U-Boot and Xen.

aaillet commented:
> I have a small comment regarding SCIF3/HSCIF0 used by Cortex A cluster. If I remember correctly, Renesas BSP moved Cortex A to HSCIF0. At least this is what we use in OP-TEE, U-Boot and Xen.

I think that we agree on this subject.

Physically SCIF3 and HSCIF0 are using the same lines and Cortex A's are using HSCIF0. (HSCIF1 and SCIF0 are on the same line as well)
Zephyr part is only supporting SCIF so I listed physically available lines in the device tree.
As typical use case is that Cortex A's are using HSCIF0, I selected SCIF0 as main Zephyr serial port.

For other use cases, Zephyr may use SCIF3 line.

Please keep in mind that HSCIF0 and SCIF0 are not on the same line, they can work at the same without any issues. (On different physical port)

I hope I answered your comments.

lorc reviewed (dismissed):
Yes, thank you.

fabiobaltieri commented on boards/arm/rcar_spider/doc/rcar_spider.rst:
```diff
@@ -0,0 +1,187 @@
+Overview
+********
+
+R-Car S4 enables the launch of Car Server/CoGW with high performance, high-speed networking, high security and high functional safety levels that are required as E/E architectures evolve into domains and zones.
```
can you wrap the lines to either 80 or 100 columns? sorry we don't have a check for this yet but that's how these docs are normally done

aaillet replied:
Ok, but how should I manage links that are longer than 100 characters then ?
They are gathered at the bottom of the rst file

fabiobaltieri replied:
those are fine, just wrap the normal text

fabiobaltieri commented on drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:
```diff
@@ -0,0 +1,208 @@
+	int ret = -ENOTSUP;
+	uint32_t reg;
+
+	enable = !!enable;
```
does not look like this is used, stale? guess you  may want to remove the argument entirely

aaillet replied:
It's not used yet as current driver don't need their clocks to be enabled or disabled manually.
This whole function is a placeholder to keep skeleton from other renesas clock driver.
It's also already planned that the arm64 support of this soc (coming just after this PR, see #60252 ) will need some clocks to be enable/disabled for some drivers.

More specifically, this argument is just as a boolean to know if this function is called to enable or disable the concerned clock.

That's all reasons why this function has to be here but is not used yet.

fabiobaltieri replied:
ok, keep the argument, drop this line? or mark it as `ARG_UNUSED`

fabiobaltieri commented on drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:
```diff
@@ -0,0 +1,208 @@
+	struct cpg_clk_info_table *clk_info;
+	struct r8a779f0_cpg_mssr_data *data = dev->data;
+	k_spinlock_key_t key;
+	int ret = 0;
```
drop the initializer, return with `return 0;` at the end of the function

fabiobaltieri commented on drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:
```diff
@@ -0,0 +1,208 @@
+	clock_control_subsys_t sys, bool enable)
+{
+	struct rcar_cpg_clk *clk = (struct rcar_cpg_clk *)sys;
+	int ret = -EINVAL;
```
no need to initialize this

aaillet replied:
was getting "may be used uninitialized" warning when declaring ret uninitialized

adding an "else" initializing ret at -EINVAL to fix warning

fabiobaltieri commented on drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:
```diff
@@ -0,0 +1,208 @@
+
+static uint32_t r8a779f0_get_div_helper(uint32_t reg_val, uint32_t module)
+{
+	uint32_t divider = RCAR_CPG_NONE;
```
how about using explicit return and ditching the variable? easier to follow...

fabiobaltieri commented on drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:
```diff
@@ -0,0 +1,208 @@
+
+static int r8a779f0_set_rate_helper(uint32_t module, uint32_t *divider, uint32_t *div_mask)
+{
+	int ret = -ENOTSUP;
```
just `return -ENOSUP` for now, you can add stuff later with the structure that makes sense at that point

aaillet commented:
Addressed comments from @fabiobaltieri concerning board documentation and clock driver.

fabiobaltieri commented on drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:
```diff
@@ -0,0 +1,202 @@
+}
+
+int r8a779f0_cpg_mssr_start_stop(const struct device *dev,
+	clock_control_subsys_t sys, bool enable)
```
sorry one more indentation nitpick: under the previous parenthesis is how it's normally done here, you are mixing thigs up a bit in the file, here you are just doing a tab, in `r8a779f0_cpg_mssr_start` you are back one charcter, could you check these and make sure they are at least coherent within the file?

```
int r8a779f0_cpg_mssr_start_stop(const struct device *dev,
                                 clock_control_subsys_t sys, bool enable)
static int r8a779f0_cpg_mssr_start(const struct device *dev,
                                   clock_control_subsys_t sys)
```

is how it should be

aaillet replied:
running clang-format on it as it is a new file could be the right call ?

fabiobaltieri replied:
clang-format mostly works but often does a bit of a disaster on complex macros, if you want to use it fine but don't take the output religiously, either commit locally, clang-format and then manually `checkout -p` the messy chunks, or add some `/* clang-format off */` `/* clang-format on */` markers around the part it messes up, if any.

fabiobaltieri commented on drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:
```diff
@@ -0,0 +1,202 @@
+	if (!ret) {
+		rcar_cpg_write(DEVICE_MMIO_GET(dev), clk_info->offset, reg);
+	}
+	return ret;
```
would not you be better off just `return -ENOTUP` and then implement the whole thing once there's something to do? I understand you want to have the stubs but it just looks really really weird right now, plus the way you are using ret around is hard to follow and does not look quite right

aaillet replied:
I just need to check if it is mandatory to write clk reg when errors are happening (l67) in the controller doc.
If it is, I'll just do this `rcar_cpg_write` before returning ENOTSUP :+1: 

Yeah, I totally understand that these placeholders are weird right now, at least keeping them with a `return -ENOTSUP` is already nice for the driver creation !

aaillet commented:
Addressed last comments from @fabiobaltieri concerning clock driver
